### DISCUSSION TITLE
read stimulus from external file

### DISCUSSION THREAD

**Discussion by sonhkim** [jspsych/jsPsych#818] [Q&A] [answered]
Hi jsPsych, 

It might be a duplicate question, but how can I read stimulus from an external file such as csv or txt? 
I have a simple survey design where participants will answer one question at a time (=per screen), and there are about 100 questions, so I can't imagine writing them into a script. Thank you!

**Comment by holgerMitterer** [comment]
May I simply second this request? I tried to find something, but was not successful.

**Comment by jswhitlo6** [comment]
Can I hop on this bandwagon and ask how this could be done for visual stimuli such as jpeg images?

**Comment by jkhartshorne** [comment, marked as answer]
So you can import stimuli, but not from a csv or txt. This has to do with browser security. What you can do is put your stimuli into a javascript file and import that. You'll want to use [JSON format](https://en.wikipedia.org/wiki/JSON), which is super simple. There are even [online tools to convert your CSV to JSON](https://csvjson.com/). 

Here's an example from one of my studies. I created a file called 'stims.js', which contains a single JSON variable called 'stims', which itself contains several arrays of stimuli:

```js
var stims = {
        pretraining  : [
                '__1-1_.mp3',
                '__1-1_.mp3'
        ],

        training  : [
          'match__1-2-3-5-4-4-4_3-5_.mp3',
          'match__3-2-1-1-3-2_3-2_.mp3'
        ],

        trials : [
          'finalfrst__1-2-3-3-3-4_1-2-3-1-1-4_.mp3',
          'finalfrst__1-2-3-5-3-2-4_1-2-3-5-4-1-4_.mp3'
        ],

        catch : [
                'catch3.mp3',
                'catch4.mp3',
                'catch5.mp3'
        ],
};
```

Then I import this into my experiment:

```html
<!doctype html>
<html>
    

    <head>
        ...
        <script src="/static/js/stims.js" type="text/javascript"></script>
        ...
    </head>
    
    <body>
    ....
```

Working with JSONs in Javascript is pretty straightforward. In fact, it's the preferred data format. So here's some code where I pull some information from the JSON:

        
```js
        var pretraining_timeline = []
        for (var stim in stims.pretraining) {
            pretraining_timeline.push({stimulus: '/static/audio/'+stims.pretraining[stim]})
        }

        

        let training1 = {
          type: 'single-audio',
          choices: ['1', '2', '3', '4', '5', '6', '7'],
          prompt: '<p>How similar are these two tones on a scale from 1 (not similar at all) to  7 (very similar).</p>',
          timeline: pretraining_timeline,
          randomize_order: true,
          //repetitions: 6
        }
```

As you can see, I can access the 'pretraining' array from my 'stims' variable that was in 'stims.js' by typing 'stims.pretraining'. And then I loop through this array and push the stimuli into my timeline. I save this (with some formatting to get the path right) to the 'pretraining_timeline' array, and then add that as a [timeline variable](https://www.jspsych.org/overview/timeline/#timeline-variables) to my jsPsych timeline.

**Reply by Quinlanptq** [reply]
Thanks very much for this I have really struggled to do a similar thing and this works just fine. Very neat.
Philip.
